gstd/src/common/handlers.rs

```diff
@@ -25,50 +25,208 @@
             }
-            _ => ext::panic("no info"),
         };
 
-        crate::debug!("panic occurred: {msg}");
-        ext::panic(&msg)
+        #[cfg(feature = "debug")]
+        let _ = ext::debug(&debug_msg);
+
+        ext::panic(&debug_msg)
+    }
+
+    /// Panic handler for stable Rust >=1.73.
+    #[cfg(all(not(feature = "panic-info-message"), feature = "panic-message"))]
+    #[panic_handler]
+    pub fn panic(panic_info: &PanicInfo) -> ! {
+        use crate::prelude::fmt::{self, Write};
+        use arrayvec::ArrayString;
+
+        #[derive(Default)]
+        struct TempBuffer<const CAP: usize> {
+            overflowed: bool,
+            buffer: ArrayString<CAP>,
+        }
+
+        impl<const CAP: usize> TempBuffer<CAP> {
+            #[inline]
+            fn write_str(&mut self, s: &str) {
+                if !self.overflowed && self.buffer.write_str(s).is_err() {
+                    self.overflowed = true;
+                }
+            }
+        }
+
+        #[derive(Default)]
+        struct TempOutput {
+            found_prefix: bool,
+            found_delimiter: bool,
+            #[cfg(feature = "panic-location")]
+            location: TempBuffer<TRIMMED_MAX_LEN>,
+            message: TempBuffer<TRIMMED_MAX_LEN>,
+        }
+
+        impl fmt::Write for TempOutput {
+            fn write_str(&mut self, s: &str) -> fmt::Result {
+                if !self.found_prefix && s.len() == PANICKED_AT.len() {
+                    self.found_prefix = true;
+                    return Ok(());
+                }
+
+                if !self.found_delimiter {
+                    if s == ":\n" {
+                        self.found_delimiter = true;
+                        return Ok(());
+                    }
+                    #[cfg(feature = "panic-location")]
+                    self.location.write_str(s);
+                } else {
+                    self.message.write_str(s);
+                }
+
+                Ok(())
+            }
+        }
+
+        let mut output = TempOutput::default();
+        let _ = write!(&mut output, "{panic_info}");
+
+        #[cfg(feature = "panic-location")]
+        let location = &*output.location.buffer;
+        let message = &*output.message.buffer;
+
+        let mut debug_msg = ArrayString::<ARRAY_STRING_MAX_LEN>::new();
+        let _ = debug_msg.try_push_str(PANIC_PREFIX);
+
+        #[cfg(feature = "panic-location")]
+        for s in ["'", message, "' at '", location, "'"] {
+            if debug_msg.try_push_str(s).is_err() {
+                break;
+            }
+        }
+
+        #[cfg(not(feature = "panic-location"))]
+        for s in ["'", message, "'"] {
+            if debug_msg.try_push_str(s).is_err() {
+                break;
+            }
+        }
+
+        #[cfg(feature = "debug")]
+        let _ = ext::debug(&debug_msg);
+
+        ext::panic(&debug_msg)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
It would be good to test every formatter

Reply:
this will require compiling wasms.
the existing test is about default rust panic message formatting (for the case of future changes as it did in 1.73)
I don't think it's that much necessary (at this moment at least). Or we can implement processing functions that we use inside different builders (`foo(&'_ PanicInfo) -> &'_ str`) and check them in unit test